CppImport/src/macro/MetaDefinitionManager.cpp

```diff
@@ -0,0 +1,249 @@
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "MetaDefinitionManager.h"
+
+#include "DefinitionManager.h"
+#include "ExpansionManager.h"
+#include "LexicalHelper.h"
+#include "StaticStuff.h"
+#include "XMacroManager.h"
+
+namespace CppImport {
+
+MetaDefinitionManager::MetaDefinitionManager(OOModel::Project* root, ClangHelper* clang,
+															DefinitionManager* definitionManager, ExpansionManager* expansionManager,
+															LexicalHelper* lexicalHelper, XMacroManager* xMacroManager)
+	: root_(root), clang_(clang), definitionManager_(definitionManager), expansionManager_(expansionManager),
+	  lexicalHelper_(lexicalHelper), xMacroManager_(xMacroManager) {}
+
+OOModel::Declaration* MetaDefinitionManager::metaDefinitionParent(const clang::MacroDirective* md)
+{
+	OOModel::Declaration* result = nullptr;
+
+	QString namespaceName, fileName;
+	if (definitionManager_->macroDefinitionLocation(md, namespaceName, fileName))
+	{
+		// find the namespace module for md
+		OOModel::Module* namespaceModule =
+				DCast<OOModel::Module>(StaticStuff::findDeclaration(root_->modules(), namespaceName));
+
+		// this assertion holds if the project structure matches Envision's project structure
+		// alternatively if no such module could be found (project structure unlike Envision's) one could put it into root_
+		Q_ASSERT(namespaceModule); //if (!namespaceModule) return root_;
+
+		// try to find the module (includes macro containers) to put this macro in
+		result = StaticStuff::findDeclaration(namespaceModule->modules(), fileName);
+
+		// if no module could be found; try to find an appropriate class to put this macro in
+		if (!result) result = StaticStuff::findDeclaration(namespaceModule->classes(), fileName);
+
+		// if no existing place could be found: create a new module (macro container) and put the macro in there
+		if (!result)
+		{
+			result = new OOModel::Module(fileName);
+			namespaceModule->modules()->append(result);
+		}
+	}
+	else
+	{
+		result = StaticStuff::findDeclaration(root_->modules(), "notenvision");
+
+		if (!result)
+		{
+			result = new OOModel::Module("notenvision");
+			root_->modules()->append(result);
+		}
+	}
+
+	Q_ASSERT(result);
+	return result;
+}
+
+OOModel::MetaDefinition* MetaDefinitionManager::metaDefinition(const clang::MacroDirective* md)
+{
+	QString h = definitionManager_->hash(md);
+
+	if (!metaDefinitions_.contains(h))
```

Comment:
Use find to avoid redundant computation. Then you won't need a separate `h` variable anymore.